### PR TITLE
ConstraintViolated event

### PR DESCRIPTION
**Pull request by dmdashenkov**
This PR introduces the `ConstraintViolated` system event.

The event is emitted whenever an attempt to transfer an entity into an invalid state occurs.

Also, this PR introduces the word `Signal` as an umbrella term for events and commands.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

The build will pass as soon as [this PR](https://github.com/SpineEventEngine/base/pull/434) is merged.

File: core/src/main/java/io/spine/core/MessageWithContext.java [hunk 1]
```diff
@@ -156,4 +158,26 @@ default boolean isBetween(Timestamp periodStart, Timestamp periodEnd) {
         TimestampTemporal end = TimestampTemporal.from(periodEnd);
         return timeTemporal.isBetween(start, end);
     }
+
+    /**
+     * Obtains the qualifier of this message.
+     *
+     * <p>A message qualifier is a digest information about the message which describes it's origin
```
Comment (armiol) [inline core/src/main/java/io/spine/core/MessageWithContext.java]: `its`.

File: core/src/main/java/io/spine/core/OriginMixin.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import static io.spine.validate.Validate.isNotDefault;
+
+/**
+ * A mixin interface for the {@link Origin} message type.
+ */
+@GeneratedMixin
+interface OriginMixin extends Message {
+
+    @SuppressWarnings("override")
+    MessageQualifier getQualifier();
```
Comment (armiol) [inline core/src/main/java/io/spine/core/OriginMixin.java]: Please document the API of this interface...

File: core/src/main/java/io/spine/core/OriginMixin.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+    default MessageQualifier qualifier() {
+        return getQualifier();
+    }
+
+    @SuppressWarnings("override")
```
Comment (armiol) [inline core/src/main/java/io/spine/core/OriginMixin.java]: ...and have a comment on this suppression.

File: core/src/main/proto/spine/core/diagnostics.proto [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import "google/protobuf/any.proto";
+
+import "spine/core/actor_context.proto";
+import "spine/core/version.proto";
+
+// An identifier of a message dispatched to an entity.
+//
+// May represent either an event or a command.
+//
+message MessageQualifier {
```
Comment (armiol) [inline core/src/main/proto/spine/core/diagnostics.proto]: Let's discuss this message and the `EntityQualifier` one. I don't think they are named according to what they are.

File: server/src/main/java/io/spine/server/entity/EntityLifecycle.java [hunk 5]
```diff
@@ -264,6 +269,26 @@ public final void onStateChanged(EntityRecordChange change,
         postIfRestored(change, dispatchedMessageIds);
     }
 
+    public final void onInvalidEntity(MessageQualifier cause,
```
Comment (armiol) [inline server/src/main/java/io/spine/server/entity/EntityLifecycle.java]: Please document.

Also `cause` parameter name seems confusing, as we also have `error` right here. They kinda interfere.

File: core/src/main/proto/spine/core/diagnostics.proto [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    option (is).java_type = "io.spine.core.MessageQualifierMixin";
+
+    google.protobuf.Any message_id = 1 [(required) = true, (validate) = true];
+
+    string message_type_url = 2 [(required) = true];
```
Comment (armiol) [inline core/src/main/proto/spine/core/diagnostics.proto]: I also think it's simply `type_url`.

File: server/src/main/proto/spine/system/server/diagnostic_events.proto [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+option java_package = "io.spine.system.server";
+option java_outer_classname = "DiagnosticEventsProto";
+option java_multiple_files = true;
+
+import "spine/core/diagnostics.proto";
+import "spine/validate/validation_error.proto";
+
+// An event emitted whenever an entity validation constraint is violated.
+//
+// None of the fields of this message is required, as we might not know something about
```
Comment (armiol) [inline server/src/main/proto/spine/system/server/diagnostic_events.proto]: Let's discuss this one as well.

**Comment by dmdashenkov** [comment]
@armiol, let's discuss.

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

**Pull request merged**